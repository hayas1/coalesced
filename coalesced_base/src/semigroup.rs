use crate::annotate::{Annotate, Annotated};

pub trait Semigroup {
    fn semigroup_op(base: Self, other: Self) -> Self;
}
pub trait AnnotatedSemigroup<A>: Sized {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
    fn default_semigroup_op(
        base: Self,
        other: Self,
        base_annotation: A,
        other_annotation: A,
    ) -> Self {
        Self::annotated_op(
            base.annotated(base_annotation),
            other.annotated(other_annotation),
        )
        .value
    }
}
impl<T: AnnotatedSemigroup<A>, A> Semigroup for Annotated<T, A> {
    fn semigroup_op(base: Self, other: Self) -> Self {
        AnnotatedSemigroup::annotated_op(base, other)
    }
}

#[cfg(any(test, feature = "test"))]
pub mod tests {
    use std::fmt::Debug;

    use crate::op::reverse::Reversed;

    use super::*;

    pub fn associative_law<T: Semigroup + Clone>(a: T, b: T, c: T) -> (T, T) {
        let ab_c = T::semigroup_op(T::semigroup_op(a.clone(), b.clone()), c.clone());
        let a_bc = T::semigroup_op(a.clone(), T::semigroup_op(b.clone(), c.clone()));
        (ab_c, a_bc)
    }
    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let (ab_c, a_bc) = associative_law(a, b, c);
        assert_eq!(ab_c, a_bc);
    }

    pub fn reverse<T: Semigroup + Clone>(a: T, b: T) -> (T, T) {
        let (ra, rb) = (Reversed(a.clone()), Reversed(b.clone()));
        (T::semigroup_op(a, b), Reversed::<T>::semigroup_op(rb, ra).0)
    }
    pub fn assert_reversed_associative_law<T: Semigroup + Clone + PartialEq + Debug>(
        a: T,
        b: T,
        c: T,
    ) {
        let (ra, rb, rc) = (Reversed(a), Reversed(b), Reversed(c));
        assert_associative_law(ra, rb, rc);
    }
}
