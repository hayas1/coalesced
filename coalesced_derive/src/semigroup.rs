use proc_macro2::TokenStream;
use syn::DeriveInput;

use crate::{constant::ConstantExt, semigroup::attr::ContainerAttr};

mod attr;
mod implementor;

pub fn gen_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;

    todo!()
}
