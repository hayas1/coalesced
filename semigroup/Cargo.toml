[package]
name = "semigroup"
version.workspace = true
description.workspace = true
authors.workspace = true
edition.workspace = true
repository.workspace = true
license-file.workspace = true
readme.workspace = true

[features]
derive = []
serde = ["dep:serde"]
test = ["monoid", "dep:rand"]
monoid = ["dep:num"]
histogram = ["dep:hdrhistogram"]

[dependencies]
semigroup_derive = { version = "0.3.2", path = "../semigroup_derive" }
serde = { version = "1.0.228", features = ["derive"], optional = true }
num = { version = "0.4.3", optional = true }
rand = { version = "0.9.2", optional = true }
hdrhistogram = { version = "7.5.4", optional = true }

[dev-dependencies]
semigroup = { path = ".", features = ["derive", "test"] }
rstest = "0.26.1"
trybuild = { version = "1.0.111", features = ["diff"] }
serde_json = "1.0.145"
serde = { version = "1.0.228", features = ["derive"] }
clap = { version = "4.5.50", features = ["derive", "env"] }


[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(doc_cfg)'] }

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "doc_cfg"]
