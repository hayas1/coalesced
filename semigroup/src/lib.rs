#![cfg_attr(doc_cfg, feature(doc_cfg))]
//! [*semigroup*](crate::semigroup::Semigroup) trait is useful for
//! - reading configs from multiple sources
//! - statistically aggregation
//! - fast range queries using segment tree
//!
//! # Usage
//! ```sh
//! cargo add semigroup --features derive,monoid
//! ```
//!
//! # Examples
//!
//! ## Reading configs from multiple sources
//! ### Simple coalesce
//! ```
//! use semigroup::Semigroup;
//! #[derive(Debug, Clone, PartialEq, Semigroup)]
//! #[semigroup(with = "semigroup::op::Coalesce")]
//! pub struct Config<'a> {
//!     pub num: Option<u32>,
//!     pub str: Option<&'a str>,
//!     #[semigroup(with = "semigroup::op::Overwrite")]
//!     pub boolean: bool,
//! }
//!
//! let cli = Config { num: Some(1), str: None, boolean: true };
//! let file = Config { num: None, str: Some("ten"), boolean: false };
//! let env = Config { num: Some(100), str: None, boolean: false };
//!
//! let config = cli.semigroup(file).semigroup(env);
//!
//! assert_eq!(config, Config { num: Some(1), str: Some("ten"), boolean: false });
//! ```
//!
//! ### Coalesce with rich enum annotation
//! Some [`Semigroup`] such as [`op::Coalesce`] can have an annotation.
//! More detail is in [`Annotate`].
//! ```
//! use semigroup::{Annotate, Semigroup};
//! #[derive(Debug, Clone, PartialEq, Semigroup)]
//! #[semigroup(annotated, with = "semigroup::op::Coalesce")]
//! pub struct Config<'a> {
//!     pub num: Option<u32>,
//!     pub str: Option<&'a str>,
//!     #[semigroup(with = "semigroup::op::Overwrite")]
//!     pub boolean: bool,
//! }
//! #[derive(Debug, Clone, PartialEq)]
//! pub enum Source {
//!     File,
//!     Env,
//!     Cli,
//! }
//!
//! let cli = Config { num: Some(1), str: None, boolean: true }.annotated(Source::Cli);
//! let file = Config { num: None, str: Some("ten"), boolean: false }.annotated(Source::File);
//! let env = Config { num: Some(100), str: None, boolean: false }.annotated(Source::Env);
//!
//! let config = cli.semigroup(file).semigroup(env);
//!
//! assert_eq!(config.value(), &Config { num: Some(1), str: Some("ten"), boolean: false });
//! assert_eq!(config.annotation().num, Source::Cli);
//! assert_eq!(config.annotation().str, Source::File);
//! assert_eq!(config.annotation().boolean, Source::Env);
//! ```
//!
//! ## Statistically aggregation
//! ### Aggregate with histogram
//! Only available with the `histogram` feature. More detail is in [`op::HdrHistogram`].
//! ```
//! # #[cfg(feature="histogram")]
//! # {
//! use semigroup::{op::HdrHistogram, Semigroup};
//!
//! let histogram1 = (1..100).collect::<HdrHistogram<u32>>();
//! let histogram2 = (100..1000).collect::<HdrHistogram<u32>>();
//!
//! let histogram = histogram1.semigroup(histogram2);
//!
//! assert_eq!(histogram.mean(), 499.9999999999999);
//! assert_eq!(histogram.value_at_quantile(0.9), 900);
//! # }
//! ```
//!
//! ### Aggregate request-response result
//! Only available with the `histogram` feature.
//! ```
//! # #[cfg(all(feature="monoid", feature="histogram"))]
//! # {
//! use std::time::{Duration, Instant};
//! use semigroup::{
//!     op::{HdrHistogram, Sum, Min, Max},
//!     Construction, Commutative, Semigroup, OptionMonoid, Monoid
//! };
//!
//! #[derive(Debug, Clone, PartialEq, Semigroup)]
//! #[semigroup(commutative)]
//! pub struct RequestAggregate {
//!     count: Sum<u64>,
//!     pass: Sum<u64>,
//!     start: Min<Instant>,
//!     end: Max<Instant>,
//!     latency: HdrHistogram<u32>,
//! }
//! impl RequestAggregate {
//!     pub fn new(pass: bool, time: Instant, latency: Duration) -> Self {
//!         Self {
//!             count: Sum(1),
//!             pass: Sum(if pass { 1 } else { 0 }),
//!             start: Min(time),
//!             end: Max(time),
//!             latency: HdrHistogram::from_iter([latency.as_millis() as u64]),
//!         }
//!     }
//!     pub fn count(&self) -> u64 {
//!         self.count.into_inner()
//!     }
//!     pub fn pass_rate(&self) -> f64 {
//!         self.pass.into_inner() as f64 / self.count.into_inner() as f64
//!     }
//!     pub fn duration(&self) -> Duration {
//!         self.end.into_inner() - self.start.into_inner()
//!     }
//!     pub fn rps(&self) -> f64 {
//!         self.count.into_inner() as f64 / self.duration().as_secs_f64()
//!     }
//!     pub fn p99_latency(&self) -> Duration {
//!         Duration::from_millis(self.latency.value_at_quantile(0.99) as u64)
//!     }
//! }
//!
//! let (now, mut agg) = (Instant::now(), OptionMonoid::unit());
//! for i in 0..10000 {
//!     let duration = Duration::from_millis(i);
//!     agg = agg.semigroup(RequestAggregate::new(i % 2 == 0, now + duration, duration).into());
//! }
//!
//! let request_aggregate = agg.into_inner().unwrap();
//! assert_eq!(request_aggregate.count(), 10000);
//! assert_eq!(request_aggregate.pass_rate(), 0.5);
//! assert_eq!(request_aggregate.duration(), Duration::from_millis(9999));
//! assert_eq!(request_aggregate.rps(), 1000.1000100010001);
//! assert_eq!(request_aggregate.p99_latency(), Duration::from_millis(9903));
//! # }
//! ```
//!
//! ## Segment tree
//! More detail is in [`segment_tree::SegmentTree`] that requires [`Monoid`].
//! ### Range sum
//! Only available with the `monoid` feature
//! ```
//! # #[cfg(feature="monoid")]
//! # {
//! use semigroup::{op::Sum, Semigroup, Construction, segment_tree::SegmentTree};
//! let data = 0..=10000;
//! let mut sum_tree: SegmentTree<_> = data.into_iter().map(Sum).collect();
//! assert_eq!(sum_tree.fold(3..6).into_inner(), 12);
//! assert_eq!(sum_tree.fold(..).into_inner(), 50005000);
//! sum_tree.update_with(4, |Sum(x)| Sum(x + 50));
//! sum_tree.update_with(9999, |Sum(x)| Sum(x + 500000));
//! assert_eq!(sum_tree.fold(3..6).into_inner(), 62);
//! assert_eq!(sum_tree.fold(..).into_inner(), 50505050);
//! # }
//! ```
//! ### Custom monoid operator
//! Only available with the `monoid` feature
//! ```
//! # #[cfg(feature="monoid")]
//! # {
//! use semigroup::{Semigroup, Construction, segment_tree::SegmentTree, Monoid};
//! #[derive(
//!     Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash, Construction,
//! )]
//! #[construction(monoid, commutative, unit = Self(i32::MIN))]
//! struct Max(pub i32);
//! impl Semigroup for Max {
//!     fn op(base: Self, other: Self) -> Self {
//!         Self(std::cmp::max(base.0, other.0))
//!     }
//! }
//!
//! let data = [2, -5, 122, -33, -12, 14, -55, 500, 3];
//! let mut max_tree: SegmentTree<_> = data.into_iter().map(Max).collect();
//! assert_eq!(max_tree.fold(3..6).0, 14);
//! max_tree.update_with(4, |Max(x)| Max(x + 1000));
//! assert_eq!(max_tree.fold(3..6).0, 988);
//!
//! // #[test]
//! semigroup::assert_monoid!(&max_tree[..]);
//! # }
//! ```
//!
//! # Links
//! - GitHub: <https://github.com/hayas1/semigroup>
//! - GitHub Pages: <https://hayas1.github.io/semigroup/semigroup>
//! - Crates.io: <https://crates.io/crates/semigroup>
//! - Docs.rs: <https://docs.rs/semigroup>
//!
//! # Testing
//! ## Benchmarks
//! // TODO
//!
//! ## Coverage
//! <https://hayas1.github.io/semigroup/semigroup/tarpaulin-report.html>
//!

mod annotate;
mod commutative;
mod construction;
mod iter;
#[cfg(feature = "monoid")]
mod monoid;
pub mod op;
#[cfg(feature = "monoid")]
pub mod segment_tree;
mod semigroup;

pub use self::{annotate::*, commutative::*, construction::*, iter::*, semigroup::*};

#[cfg(feature = "monoid")]
pub use self::monoid::*;

#[cfg(feature = "derive")]
#[cfg_attr(doc_cfg, doc(cfg(feature = "derive")))]
pub use semigroup_derive::{properties, Construction, Semigroup};
