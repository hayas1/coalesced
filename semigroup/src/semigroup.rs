use crate::Annotated;

/// [`Semigroup`] represents a binary operation that satisfies the following properties
/// 1. *Closure*: `op: T × T → T`
/// 2. *Associativity*: `op(op(a, b), c) = op(a, op(b, c))`
///
/// # Examples
/// ## Deriving
/// When fields do not implement [`Semigroup`], use `with` attribute.
/// ```
/// use semigroup::Semigroup;
/// #[derive(Debug, Clone, PartialEq, Semigroup)]
/// #[semigroup(with = "semigroup::op::Coalesce")]
/// pub struct ExampleStruct<'a> {
///     pub str: Option<&'a str>,
///     #[semigroup(with = "semigroup::op::Overwrite")]
///     pub boolean: bool,
///     #[semigroup(with = "semigroup::op::Sum")]
///     pub sum: u32,
/// }
///
/// let a = ExampleStruct { str: None, boolean: true, sum: 1 };
/// let b = ExampleStruct { str: Some("ten"), boolean: false, sum: 10 };
/// let c = ExampleStruct { str: None, boolean: false, sum: 100 };
///
/// // #[test]
/// semigroup::assert_semigroup!(&a, &b, &c);
/// assert_eq!(a.semigroup(b).semigroup(c), ExampleStruct { str: Some("ten"), boolean: false, sum: 111 });
/// ```
///
/// ## Construction
/// [`Semigroup`] can be constructed by [`crate::Construction`].
///
/// Some operations are already provided by [`crate::op`].
/// ```
/// use semigroup::{Construction, Semigroup};
///
/// #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash, Construction)]
/// pub struct Sum(u64);
/// impl Semigroup for Sum {
///     fn op(base: Self, other: Self) -> Self {
///         Self(base.0 + other.0)
///     }
/// }
///
/// let (a, b, c) = (Sum(1), Sum(2), Sum(3));
/// // #[test]
/// semigroup::assert_semigroup!(&a, &b, &c);
/// assert_eq!(a.semigroup(b).semigroup(c), Sum(6));
/// ```
///
/// # Testing
/// Use [`crate::assert_semigroup!`] macro.
///
/// The *closure* property is guaranteed by Rust’s type system,
/// but *associativity* must be verified manually using [`crate::assert_semigroup!`].
pub trait Semigroup {
    fn op(base: Self, other: Self) -> Self;
    fn semigroup(self, other: Self) -> Self
    where
        Self: Sized,
    {
        Semigroup::op(self, other)
    }
}

/// [`AnnotatedSemigroup`] is a [`Semigroup`] that has an annotation, such as [`crate::Annotate`].
pub trait AnnotatedSemigroup<A>: Sized + Semigroup {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
}

#[cfg(feature = "test")]
pub mod test_semigroup {
    use std::fmt::Debug;

    use rand::seq::IndexedRandom;

    use crate::{
        combine::test_combine::{assert_combine_iter, assert_lazy},
        commutative::test_commutative::{assert_reverse_associative_law, assert_reverse_reverse},
    };

    use super::*;

    /// Assert that the given type satisfies the *semigroup* property.
    ///
    /// # Usage
    /// - 1 argument: iterator of more than 3 items that implements [`Semigroup`].
    /// - More than 3 arguments: items that implements [`Semigroup`].
    ///
    /// # Examples
    /// ```
    /// use semigroup::{assert_semigroup, op::Coalesce};
    ///
    /// let a = Coalesce(Some(1));
    /// let b = Coalesce(None);
    /// let c = Coalesce(Some(3));
    /// assert_semigroup!(a, b, c);
    ///
    /// let v = vec![a, b, c];
    /// assert_semigroup!(&v);
    /// ```
    ///
    /// # Panics
    /// - If the given function does not satisfy the *semigroup* property.
    /// - The input iterator has less than 3 items.
    ///
    /// ```compile_fail
    /// use semigroup::{assert_semigroup, op::Coalesce};
    /// let a = Coalesce(Some(1));
    /// let b = Coalesce(None);
    /// assert_semigroup!(a, b);
    /// ```
    #[macro_export]
    macro_rules! assert_semigroup {
        ($a:expr, $b: expr, $($tail: expr),*) => {
            {
                let v = vec![$a, $b, $($tail),*];
                $crate::assert_semigroup!(&v)
            }
        };
        ($v:expr) => {
            {
                let (a, b, c) = $crate::test_semigroup::pick3($v);
                $crate::test_semigroup::assert_semigroup_impl(a.clone(), b.clone(), c.clone());
            }
        };
    }

    pub fn pick3<T: Clone>(data: &[T]) -> (T, T, T) {
        data.choose_multiple_array(&mut rand::rng())
            .map(|[a, b, c]| (a, b, c))
            .expect("failed to pick 3 items")
    }

    pub fn assert_semigroup_impl<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        assert_associative_law(a.clone(), b.clone(), c.clone());
        assert_reverse_reverse(a.clone(), b.clone(), c.clone());
        assert_reverse_associative_law(a.clone(), b.clone(), c.clone());
        assert_combine_iter(a.clone(), b.clone(), c.clone());
        assert_lazy(a.clone(), b.clone(), c.clone());
        #[cfg(feature = "monoid")]
        crate::test_monoid::assert_option_monoid(a.clone(), b.clone(), c.clone());
    }

    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let ab_c = Semigroup::op(Semigroup::op(a.clone(), b.clone()), c.clone());
        let a_bc = Semigroup::op(a.clone(), Semigroup::op(b.clone(), c.clone()));
        assert_eq!(ab_c, a_bc);
    }
}
