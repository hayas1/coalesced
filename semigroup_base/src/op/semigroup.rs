#[cfg(feature = "monoid")]
pub mod gcd;
#[cfg(feature = "monoid")]
pub mod lcm;
pub mod max;
pub mod min;
pub mod prod;
pub mod sum;
pub mod xor;
