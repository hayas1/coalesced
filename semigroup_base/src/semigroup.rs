use crate::annotate::Annotated;

pub trait Semigroup {
    fn semigroup_op(base: Self, other: Self) -> Self;
    fn semigroup(self, other: Self) -> Self
    where
        Self: Sized,
    {
        Semigroup::semigroup_op(self, other)
    }
}
pub trait AnnotatedSemigroup<A>: Sized + Semigroup {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
}

#[cfg(any(test, feature = "test"))]
pub mod tests {
    use std::fmt::Debug;

    use crate::{
        iter::tests::assert_lazy_evaluation_iter,
        reverse::tests::{assert_reverse, assert_reverse_associative_law},
    };

    use super::*;

    #[macro_export]
    macro_rules! assert_semigroup_op {
        ($a:expr, $b: expr, $c: expr) => {
            $crate::semigroup::tests::assert_semigroup_op_impl($a.clone(), $b.clone(), $c.clone());
            $crate::monoid::tests::assert_option_monoid($a.clone(), $b.clone(), $c.clone());
        };
    }
    pub use assert_semigroup_op;

    pub fn assert_semigroup_op_impl<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        assert_associative_law(a.clone(), b.clone(), c.clone());
        assert_reverse(a.clone(), b.clone(), c.clone());
        assert_reverse_associative_law(a.clone(), b.clone(), c.clone());
        assert_lazy_evaluation_iter(a.clone(), b.clone(), c.clone());
    }

    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let ab_c = T::semigroup_op(T::semigroup_op(a.clone(), b.clone()), c.clone());
        let a_bc = T::semigroup_op(a.clone(), T::semigroup_op(b.clone(), c.clone()));
        assert_eq!(ab_c, a_bc);
    }
}
