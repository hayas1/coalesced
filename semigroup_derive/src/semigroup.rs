use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::DeriveInput;

use crate::{
    constant::ConstantExt,
    semigroup::{ast::Semigroup, attr::ContainerAttr},
};

mod ast;
mod attr;

pub fn impl_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;
    let semigroup = Semigroup::new(&constant, derive, &attr)?;
    Ok(semigroup.into_token_stream())
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use crate::constant::{External, Internal};

    use super::*;

    #[rstest]
    #[case::semigroup_annotated(
        "semigroup_annotated",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated)]
            pub struct NamedStruct {
                #[semigroup(with = "semigroup::op::Overwrite")]
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: semigroup::op::Overwrite<bool>,
            }
        },
    )]
    #[case::semigroup_not_annotated(
        "semigroup_not_annotated",
        impl_semigroup::<Internal>,
        syn::parse_quote! {
            #[derive(SemigroupPriv)]
            #[semigroup(with = "semigroup::op::Overwrite")]
            pub struct UnnamedStruct<T: std::ops::Add> (
                #[semigroup(with = "semigroup::op::Added")]
                T,
                u64
            );
        },
    )]
    #[case::semigroup_custom_annotation(
        "semigroup_custom_annotation",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated, annotation_param = X, with = "semigroup::op::Overwrite")]
            pub struct NamedStruct{
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: bool,
            }
        },
    )]
    #[case::semigroup_monoid(
        "semigroup_monoid",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(monoid, commutative)]
            pub struct MonoidStruct{
                #[semigroup(with = "semigroup::op::Sum")]
                pub sum: u32,
            }
        },
    )]
    fn test_derive_semigroup_snapshot(
        #[case] case: &str,
        #[case] f: impl Fn(&DeriveInput) -> syn::Result<TokenStream>,
        #[case] input: DeriveInput,
    ) {
        let generated = f(&input).unwrap();
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_snapshot!(case, formatted);
        });
    }
}
