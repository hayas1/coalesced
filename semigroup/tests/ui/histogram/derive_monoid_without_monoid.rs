use semigroup::Semigroup;

#[derive(Debug, Clone, PartialEq, Semigroup)]
#[semigroup(monoid, with = "semigroup::op::Coalesce")]
pub struct NamedStruct {
    pub num: Option<u32>,
    #[semigroup(with = "semigroup::op::Overwrite")]
    pub boolean: bool,
}

fn main() {}
