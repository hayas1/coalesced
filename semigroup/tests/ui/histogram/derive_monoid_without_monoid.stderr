error[E0277]: the trait bound `Overwrite<_>: ConstructionMonoid<_>` is not satisfied
 --> tests/ui/histogram/derive_monoid_without_monoid.rs:3:35
  |
3 | #[derive(Debug, Clone, PartialEq, Semigroup)]
  |                                   ^^^^^^^^^ the trait `ConstructionMonoid<_>` is not implemented for `Overwrite<_>`
  |
  = help: the following other types implement trait `ConstructionMonoid<T>`:
            `Coalesce<T>` implements `ConstructionMonoid<Option<T>>`
            `Gcd<T>` implements `ConstructionMonoid<T>`
            `HdrHistogram<T>` implements `ConstructionMonoid<hdrhistogram::Histogram<T>>`
            `Lcm<T>` implements `ConstructionMonoid<T>`
            `Max<T>` implements `ConstructionMonoid<T>`
            `Min<T>` implements `ConstructionMonoid<T>`
            `OptionMonoid<T>` implements `ConstructionMonoid<Option<T>>`
            `Prod<T>` implements `ConstructionMonoid<T>`
          and $N others
  = note: this error originates in the derive macro `Semigroup` (in Nightly builds, run with -Z macro-backtrace for more info)
